core/src/main/java/com/linecorp/armeria/common/multipart/MultipartEncoder.java

```diff
@@ -0,0 +1,305 @@
+
+    private static final AtomicIntegerFieldUpdater<MultipartEncoder> subscribedUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(MultipartEncoder.class, "subscribed");
+
+    private static final HttpData CRLF = HttpData.ofUtf8("\r\n");
+
+    static final SubscriptionOption[] EMPTY_OPTIONS = {};
+
+    private final String boundary;
+
+    private final StreamMessage<BodyPart> publisher;
+
+    private volatile int subscribed;
+
+    // 'closeCause' will be written before 'closed' and read after 'closed'
+    @Nullable
+    private Throwable closeCause;
+    private volatile boolean closed;
+
+    @Nullable
+    private volatile CompletableFuture<Void> completionFuture;
+
+    @Nullable
+    private volatile DefaultStreamMessage<StreamMessage<HttpData>> emitter;
+
+    MultipartEncoder(StreamMessage<BodyPart> publisher, String boundary) {
+        requireNonNull(boundary, "boundary");
+        requireNonNull(publisher, "publisher");
+
+        this.boundary = boundary;
+        this.publisher = publisher;
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        final StreamMessage<StreamMessage<HttpData>> emitter = this.emitter;
+        if (emitter != null) {
+            return emitter.whenComplete();
+        }
+
+        final CompletableFuture<Void> completionFuture = new CompletableFuture<>();
+        if (completionFutureUpdater.compareAndSet(this, null, completionFuture)) {
+            return completionFuture;
+        } else {
+            return this.completionFuture;
+        }
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor) {
+        subscribe(subscriber, executor, EMPTY_OPTIONS);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only one Subscriber allowed"));
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        publisher.subscribe(new BodyPartSubscriber(subscriber, executor, options), executor, options);
+    }
+
+    @Override
+    public void abort() {
+        abort(AbortedStreamException.get());
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        if (closed) {
```

Comment:
How about making `closed` as non-volatile?
We can only access the variable only in the event loop.

Reply:
That makes sense.